pootle/apps/virtualfolder/receivers.py

```diff
@@ -0,0 +1,60 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+from pootle_store.models import Unit
+
+from virtualfolder.models import VirtualFolder
+
+
+@receiver(post_save, sender=VirtualFolder)
+def vfolder_unit_priority_handler(sender, **kwargs):
+
+    for unit in kwargs["instance"].units.all():
+        unit.set_priority()
+
+
+@receiver(post_save, sender=Unit)
+def relate_unit(sender, instance, created=False, **kwargs):
+    """Add newly created units to the virtual folders they belong, if any.
+
+    When a new store or translation project, or even a full project is added,
+    some of their units might be matched by the filters of any of the
+    previously existing virtual folders, so this signal handler relates those
+    new units to the virtual folders they belong to, if any.
+    """
+    if not created:
+        instance.set_priority()
```

Comment:
Unit priority must be set this in this handler only when it is created. Already existing units already have the right priority because post_save handler for VirtualFolder already set it. But new units might be added to a store included in a virtual folder, and thus its priority must be calculated here.